### ISSUE TITLE
Proper physical treatment of sound waves for improved spatial audio

### ISSUE THREAD

**Issue by Krzmbrzl** [mumble-voip/mumble#6597]
### Context

Spatial audio (also called positional audio) in Mumble uses a somewhat simplistic approach to simulating how an observer receives sound from different locations etc.

### Description

In order to maximize the realism of positional audio, Mumble should properly take physical effects into account. These are
- Use of proper [head-related transfer functions](https://en.wikipedia.org/wiki/Head-related_transfer_function) (HRTFs). This simulates how sound waves are affected by traveling through the physical object that is the human being or more specifically, its head, of the person hearing that sound. This includes frequency changes (which are themselves frequency-dependent) as well as time- and phase- and volume-shifts of the sound waves. In other words, this also incorporates interaural delay.
- Properly account for doppler-shifts arising from the (relative) movements of the sound source and the sound receiver.
- Take the virtual surroundings into account. This of course requires knowledge of some kind of 3D world map in order to be able to map from positions to physical surroundings. This would allow to have effects for
    - Reverb
    - Occlusion
    - Attenuation

HRTF could be achieved by making use of the OpenAL (open audio library) ecosystem of which there exists an LGPL licensed variant called OpenAL-Soft.
Potentially, OpenAL-Soft could also be used as a regular cross-platform audio library which would then make our own platform-specific backend implementations obsolete (which would greatly reduce the maintenance effort). Whether or not this is a viable direction is not yet clear though.

The speed of entities required for Doppler shifts can be obtained via "numeric differentiation", i.e. taking two position updates and then checking how much the entity has moved in the given amount of time.

For the environmental effects, data about the physical surroundings of the entities is required. In order to obtain these, the plugin API could be extended to allow plugins to provide this kind of data.

Note that there exists also a [Qt component](https://doc.qt.io/qt-6/qtspatialaudio-index.html) for handling spatial audio that might be a viable option for us to use. However, ideally we wouldn't introduce a Qt dependency on the audio level processing.

### Mumble component

Client

### OS-specific?

No

### Additional information

Online resources:
- https://en.wikipedia.org/wiki/Head-related_transfer_function
- https://en.wikipedia.org/wiki/Sound_localization
- https://ieeexplore.ieee.org/document/5661988
- https://github.com/kcat/openal-soft
- https://www.openal.org/
- https://github.com/layeh/barnard (Mumble client for the terminal that makes use of OpenAL)
- https://doc.qt.io/qt-6/qtspatialaudio-index.html

Related issues:
- #6532 
- #2324
- #5934
- #1933
- #3234

**Comment by Hiradur** [comment]
> The speed of entities required for Doppler shifts can be obtained via "numeric differentiation", i.e. taking two position updates and then checking how much the entity has moved in the given amount of time.

Please note that Mumble would have to know the ingame unit in which the movement speed is measured. This could be a parameter that a game-plugin could set. But even then I think It could cause some weird artifacts, e.g. if a player teleports from one end of a map to another (huge difference between positions in a singe time step). Some upper limit to safe guard around this would make sense.

> For the environmental effects, data about the physical surroundings of the entities is required. In order to obtain these, the plugin API could be extended to allow plugins to provide this kind of data.

This would be one way to do it, Creative chose another for [EAX Voice](https://web.archive.org/web/20070601013325/http://www.soundblaster.com/eax/abouteax/eax5ahd/eax5_2.asp): back when EAX was popular, it received the environmental data from the game through the sound card driver and applied that to the microphone input stream so that the processed stream was available to any VoIP software.
I don't think that OpenAL Soft supports this at the moment and it would only work for games using EFX or EAX provided by OpenAL but it wouldn't require any work on Mumble's side.

Here are some examples of EAX Voice:
https://www.youtube.com/watch?v=30fTc5t5QNU
https://www.youtube.com/watch?v=wxIYNG4TQ7U

**Comment by Krzmbrzl** [comment]
> Mumble would have to know the ingame unit in which the movement speed is measured.

I would argue that the already require the positional data to be in meters and since the respective audio is realtime, it would make sense for the time to be measured in seconds as well.

In order to account for games with very fast movement (e.g. cars or even spaceships) the plugin could set a speed multiplier in order to keep the Doppler effect on a sane level.

> Some upper limit to safe guard around this would make sense.

Absolutely!

> Creative chose another for EAX Voice

Interesting approach. Never heard of it. It sounds very convenient though.

**Comment by davidebeatrici** [comment]
I was aware of EAX, but not EAX Voice. That feature is/was cool!

I already had a technique like that in mind, but the issue (as usual) is supporting specific games. In theory we could gather data directly from the audio library if a known/documented one is used, but otherwise it's going to be hard unless somebody has already reverse engineered the internals.

**Comment by QmwJlHuSg9pa** [comment]
Your best bet would probably be to speak to the maintainer of openal-soft directly; kcat has made strides in recent years into integrating EAX support into the project.

**Comment by mirh** [comment]
> This would be one way to do it, Creative chose another for EAX Voice

I mean.. that's just a matter of different "places" where the mic effects are implement/offer the mic effects . But game-side there is no difference into a "predisposition" being required. 

And in this sense, while openal integration could certainly smooth out things for the games using it, I'm somewhat worried that the others with [some](https://learn.microsoft.com/en-us/windows/win32/xaudio2/how-to--integrate-x3daudio-with-xaudio2)/[degree](https://learn.microsoft.com/en-us/windows/win32/api/xaudio2fx/nf-xaudio2fx-xaudio2createreverb#remarks) of generalization may instead be penalized by going higher level (though openal could still be super useful to implement HRTF and whatnot spatial)

> but it wouldn't require any work on Mumble's side.

That sound card driver thing by Creative? Of course not, it works for everybody. But we don't control RTKVHD64 or AtihdWT6.
So, either you find a way to implement this in an APO (I'm not even sure it is possible, given that they would still have to poke inside game processes) or openal will have to expose this information to the rest of the system some way. 

Such ~frontend conundrum that would then stack with the one I was left with for the backend at https://github.com/kcat/openal-soft/issues/415#issuecomment-2308399677

> kcat has made strides in recent years towards integrating EAX support into the project.

EAX \*has\* been integrated, nearly 3 years ago in one big PR already.

**Comment by will-ca** [comment]
> I would argue that the already require the positional data to be in meters and since the respective audio is realtime, it would make sense for the time to be measured in seconds as well.

Would it make sense for distance/position be unitless, and instead allow specifying a speed of sound parameter? (Default `340s^-1`, equivalent to meters at STP.)

Though IDK if that'll affect accounting for wavelength.

**Comment by Krzmbrzl** [comment]
> Would it make sense for distance/position be unitless, and instead allow specifying a speed of sound parameter? (Default 340s^-1, equivalent to meters at STP.)

Not sure what problem this would try to solve though 🤔

**Comment by hiinaspace** [comment]
I slopped up a fairly minimal diff for implementing HRTF spatialization, viewable in https://github.com/hiinaspace/mumble/pull/1 . I think just HRTF evaluation alone is doable without depending on a whole library like openAL (or steam audio). An HRTF is essentially a bunch of recorded impulse responses on spherical coordinates around a head, so to use it, you just need a parser/interpolator (libmysofa) and a convolution with the source audio (through FFT).

If it's useful I can open a PR on this actual repository, but given it's all AI-generated I figured I'd comment here first.

**Comment by davidebeatrici** [comment]
Please open the PR, we'll take it from there!

**Comment by Hiradur** [comment]
While I welcome the initiative, I'd like to point out that OpenAL support would bring more than just plain HRTF support:

* optimized processing on various CPU archictectures using intrinsics
* various effects, such as
  * reverb presets (for inspiration how it could be used, see EAX Voice above)
  * doppler shift
  * obstruction and occlusion
* 3D positional audio for various speaker configurations, from stereo over 5.1, 7.1 to any arbitrary speaker configuration via Ambisonics

**Comment by hiinaspace** [comment]
Fair point. I did also consider linking mumble to all of https://github.com/ValveSoftware/steam-audio, which also does a bunch of fancy raytraced (optionally on the GPU) reverb occlusion and ambisonic/multi-speaker rendering. It's used in games like HL: Alyx and (I think) CS2, and is apache2-licensed now.

However, actually getting enough of the game state into mumble to do meaningful reverb/occlusion seems difficult. You'd have to share some sort of level geometry such that steam-audio or OpenAL can raycast into it, and I'm not aware of any existing solutions that do that cleanly. There is maybe some half-measure where e.g. the game sends some coarse reverb parameters like "player is in a 10x5x10 room with wooden walls" along with the positions using shared memory, and then mumble could do reverb from that (IIRC the now-dead [Oculus Spatializer](https://developers.meta.com/horizon/documentation/unity/audio-osp-unity-req-setup/) did something where it sampled raycasts into the game geometry to dynamically resize parametric reverb boxes). But for occlusion, you'd need to do the audio ray-tracing between players in the game and share that to mumble, or share the geometry itself (or some spatial acceleration structure). 

Meanwhile, filtering through an HRTF essentially does what mumble's headphone spatialization does today (interaural delay/attenuation) with the same input (angle between source and listener), but it sounds way better. So I still think it's a useful incremental improvement that doesn't need to change the server/link plugins at all or invent a shared geometry API.

Now that I'm thinking about this again, a gentle ping to the mumble devs that I'd still like feedback on the general approach in #7085 . And if you're a windows mumble user, you can test out the HRTF spatialization with my (unofficial) binary at https://github.com/hiinaspace/mumble/releases/tag/hrtf-experimental-0.1 . Should work with existing servers/link plugins.

**Comment by Krzmbrzl** [comment]
> gentle ping to the mumble devs that I'd still like feedback on the general approach

I'm afraid that at least from my side that won't happen soon. I have about 50 items in my inbox that require my attention and in the grand scheme of things HRTF currently doesn't have highest priority 👀

**Comment by mirh** [comment]
Ambisonics would be a nice addition and first step, though if I start to think bigger about the needs of tracing a few worries come up.
Like, obviously this doesn't matter if we are talking about plugin-supported games.. but I feel like that even if there were no concerns for overhead, sharing even the level geometry through an api could start to be too much "transparent openness" of an architecture for the tastes of many competitive games? And so a bit like EAX voice, I believe the native Mumble Link should be expanded to have the processing handled in-game (that would also have the nice advantage that you aren't duplicating representations that they are already doing for their own purposes, or becoming crazy with even having to handle materials/reverb/whatnot).

**Comment by hiinaspace** [comment]
Hmm, do you have something specific in mind for ambisonics? From my experience, if you already have separate mono point audio sources (like in voice chat), interpreting audio as spherical harmonics doesn't give you anything extra. If you're rendering to headphones/binaural, you can go straight to stereo through an HRTF. And if you're rendering to multiple speakers, you can also pretty directly attenuate per speaker. The most real thing I think "using ambisonics" translates into is if you do server-side mixing of multiple speakers, where it would then be mildly useful to mix >4 voices into 4-channel ambisonic soundfields.

More generally, I think a axis of tension with mumble doing "full physical audio simulation" is that mumble is primarily meant to transmit voices that you can understand. mono channels piped directly into your headphones/speakers optimizes that. and even within a game, you generally still want that (then justify it in the fiction with radios/walkie-talkies or magic telepathy, etc). Doing extra reverb/attenuation/occlusion/filtering almost always makes the voices less understandable. So I get why doing this is low priority for the mumble devs. There is a small argument to be made that spatialization does improve understanding of specific speakers in crowded channels, aka the cocktail party effect, but practically I think it's minimal.

The other related tension is how much a mumble plugin or mumble link-style API really adds over a multiplayer game just doing voice transmission and playback internally and integrating with openAL/steamaudio itself. The game likely already has its own audio subsystem for other sound effects anyway. And you also don't have to awkwardly map the game's teams or other player state like deadchat onto mumble's permissions and rooms. And this applies both to the actual game devs and mod devs (where the mod devs might be the ones otherwise implementing a mumble-link mod or plugin).

Practical example of this in Minecraft, where while there is a MumbleLink mod https://github.com/zsawyer/MumbleLink , nowdays people use https://github.com/henkelmax/simple-voice-chat/,  which does its own opus-over-UDP transmission, its own team/channel/shout control layer, OpenAL for spatialization, and you can also use it with https://modrinth.com/mod/sound-physics-remastered to do all the fancy reverb/attenuation/occlusion on geometry you want.

On the other hand, doing your own opus-over-UDP transmission isn't like entirely trivial. So as you say there maybe is still room for mumble link's API to still handle the input/networking/jitter/playout buffering, but then hand the game a PCM audio ring buffer per player and let the game spatialize/play them back however it wants. This is closer to what livekit (a fairly new webrtc-based client/server libary) gives you in e.g. https://github.com/NodotProject/godot-livekit . It handles encoding/transmission/broadcast, but the game engine does the final mix. One advantage that style of API could have is being able to fallback to mumble-client's own audio output if the game stops responding. One nice thing I've noticed in my own use of mumble is that even if the multiplayer game I'm trying to play is buggy/crashes the lobby, you can still talk to your friends and tell them the new lobby code/debug stuff over voice (vs if the game is doing voice, then you lose all comms when it crashes).

tl;dr; I think voice chat physical simulation is cool, but maybe it actually isn't worth doing in mumble client directly.

**Comment by mirh** [comment]
> Hmm, do you have something specific in mind for ambisonics?

Nope, sorry. I think I was mixing up https://github.com/kcat/openal-soft/issues/999 with https://github.com/kcat/openal-soft/issues/586.

> how much a mumble plugin or mumble link-style API really adds over a multiplayer game just doing voice transmission and playback internally and integrating with openAL/steamaudio itself

That would be the neat part I think, it doesn't have to be a choice.
Just like, say, openal can abstract spatialization (and whatever [library](https://github.com/vercidium-patreon/godot_raytraced_audio) can in turn build on its top for further abstractions).. whatever mumble link 2.0 could help abstracting the "player-to-player interactions".

Though now that I think better to it, it's a bit unclear just how much this kind of sdk should include.
Because if it's too barebone then as you were saying idk how much the effort could be worth, if nonetheless game devs are already doing 100% of the work themselves (yeah maybe some players might like some degree of continuity with their friends before-and-after running the game, but unlike mumble link 1.0 you aren't gaining extra cool sound features).
On the other hand something big that takes care of all voice communication itself could provide much developer value (and it would be great for games to use libxound rather than whatever crappy high latency thingy they ships normally) but it would be a much more complex and held to higher standards, and probably an offshot of libmumble directly in competition with Vivox.

> tl;dr; I think voice chat physical simulation is cool, but maybe it actually isn't worth doing in mumble client directly.

That, I agree.
This would mean then that while you are connected to a supported game *that* is passing the processed voices to you.
..which would mean you are dependent on its audio engine, for the output if not very likely even for the inputs?
I admit this starts to feel very hairy.

**Comment by Krzmbrzl** [comment]
FYI: the new plugin architecture (Mumble 1.4+) already allows plugins to manipulate audio streams however they please. So if a (game) dev wanted, they could already do all the spatialization themselves instead of providing Mumble with positional data. In fact, a use case such as that was the original motivation for the new plugin system. However, that won't work via the Link plugin as that's not designed to handle this

**Comment by Hiradur** [comment]
> Hmm, do you have something specific in mind for ambisonics?

I was mostly thinking about supporting arbitrary speaker configurations and the improved listening experience over pan-pot for surround systems since [the latter doesn't work as well, especially to the sides](https://www.ambisonic.net/ambidvd.html):

> But there are other problems with using level as the sole means of localization. The human ear/brain combination, in contrast, uses a number of different localization techniques at different, overlapping frequencies, which are at least as important. We use phase to localize sounds between 150 Hz and 1.5 kHz, while level is utilized between 300 Hz and 5 kHz. Above 2.5 kHz, other directional cues are used. Luckily, when two speakers in front of you are separated by no more than 60 degrees, both ears hear both speakers and LF amplitude differences are converted into phase differences between the ears. However, this effect works only poorly if the speakers are behind you, and not at all if the speakers are to the side. This means that, by definition, traditional quad – and today’s 5.1 systems which are their descendants – cannot work. As Martin Leese puts it in his invaluable Ambisonic FAQ, "…Any surround system that relies on pairwise mixing between adjacent speakers must fail."

With regard to environmental audio: Maybe the Mumble Link plugin could offer an [I3DL2](http://gameaudioworkshop.org/wp-content/uploads/2012/06/3dl2v1a.pdf) compatible interface that is then mapped to OpenAL EFX (or wait for [a native implementation](https://github.com/kcat/dsoal/issues/107)) or potentially another backend's reverb parameters (e.g. [XAudio2](https://learn.microsoft.com/en-us/windows/win32/api/xaudio2fx/ns-xaudio2fx-xaudio2fx_reverb_i3dl2_parameters)).
This would establish a bridge between a game's internal reverb implementation and the reverb implementation that Mumble uses. However, game and Mumble then might produce differently sounding reverb even with the same parameters if the reverb implementations differ.

**Comment by hiinaspace** [comment]
>FYI: the new plugin architecture (Mumble 1.4+) already allows plugins to manipulate audio streams however they please. So if a (game) dev wanted, they could already do all the spatialization themselves instead of providing Mumble with positional data.

Oh cool, this is `mumble_onAudioSourceFetched`, right?
https://github.com/mumble-voip/mumble/blob/9c2e988659dfad17f7eed5b0c2ffc5d10c8670be/plugins/MumblePlugin.h#L1018-L1037

Thinking about how you'd use this in the "game takes over audio output from mumble while active", I guess you'd copy the buffer into the game's own audio subsystem, then zero the buffer and return true, so mumble's own audio output is muted. Then if/when the game stops responding (exit or crash), mumble takes over again. I'll have to give it a try.

---

>I was mostly thinking about supporting arbitrary speaker configurations and the improved listening experience over pan-pot for surround systems since [the latter doesn't work as well, especially to the sides](https://www.ambisonic.net/ambidvd.html):

Ok, I think I understand the difference between simple pan-pot vs inter-speaker delays/phasing, though I still don't understand how people drive arbitrary speaker configurations in practice. Are there sound cards/hardware that accepts ambisonic channels as input? I'm vaguely aware that Dolby Atmos instead expects you to encode point sources with relative position (e.g. individual voices) vs ambisonic fields. I'm also skeptical people actually uses arbitrary multi-speaker setups with mumble voice chat in the first place. It seems like it'd be really hard to do echo cancellation. 

So perhaps more productively, @mirh @Hiradur do you have a specific game/speaker setup in mind for using mumble with full physical attenuation/reverb/occlusion/etc? I know enough about mumble's build setup now that I can probably maintain like a `spatial-mumble` fork with these features, while we're waiting for upstream. It'd only be worth it if you could actually test the features though. For #7085 , I'm testing with https://github.com/hiinaspace/ovr_mumble_link and it definitely improves mumble with VR headsets (6DOF head tracking makes it a lot more obvious when things aren't spatialized well). But I don't have a multispeaker setup, or a game in mind that would benefit from attenuation/reverb/occlusion that doesn't also have its own voice chat already.

**Comment by mirh** [comment]
>  and the reverb implementation that Mumble uses. 

There should be no such a thing, which would be a can of worms I cannot even imagine the bottom of.
It's one thing to assume people sitting at different angles and distances in an open field, it's another to imply an entire game engine.
(it's already lowbrow enough with EAX reverb and its dozens fixed presets when a developer bothered to manually implement the sound environments in their levels, but mumble without game integration would be at best as useless as the setting in the realtek control panel)

> So perhaps more productively, @mirh 

As I said I was mixing up ambisonics with the spatial audio api that could work with atmos (whose equipment I don't have, it was just the most fancy thing I could think of)

p.s. a random google made me find that flightgear got a [fully fledged client](https://gitlab.com/flightgear/flightgear/-/work_items/3095) integration (which isn't really useful if we are working with the plugin api, but maybe some concepts could be interesting)

**Comment by Krzmbrzl** [comment]
> Oh cool, this is mumble_onAudioSourceFetched, right?

Among others, yes.

**Comment by Hiradur** [comment]
> There should be no such a thing, which would be a can of worms I cannot even imagine the bottom of.
It's one thing to assume people sitting at different angles and distances in an open field, it's another to imply an entire game engine.
(it's already lowbrow enough with EAX reverb and its dozens fixed presets when a developer bothered to manually implement the sound environments in their levels, but mumble without game integration would be at best as useless as the setting in the realtek control panel)

I think you misunderstood me. My idea was to add an I3DL2 interface to the Link plugin for game integration so that the current reverb parameters of a game could be set for Mumble via this interface.

> Ok, I think I understand the difference between simple pan-pot vs inter-speaker delays/phasing, though I still don't understand how people drive arbitrary speaker configurations in practice.

You bring up a good point, driving speaker configurations beyond 7.1 is difficult but not impossible as of now.
See e.g. this tutorial for using OpenAL with Ambisonics rendering to output to such systems [here](https://gist.github.com/Hiradur/388cb7f658fe117a1f4ccfd9a21adffa) and a discussion for how Pipewire could drive such systems [here](https://gitlab.freedesktop.org/pipewire/pipewire/-/work_items/2310).

> I'm also skeptical people actually uses arbitrary multi-speaker setups with mumble voice chat in the first place. It seems like it'd be really hard to do echo cancellation.

I have a 4.1.2 system and sometimes use it with a microphone for communication. I think microphone directivity is important for this use case.

The combination of a surround speaker system + headset has been fascinating me since I made initial contact with it on XBOX 360 and XBOX Live.
The [original XBOX 360 headsets consist of a microphone and a single ear-piece](https://support.xbox.com/en-US/help/xbox-360/accessories/connecting-headsets) and are only used for voice communication but not ingame audio. So I believe, speaker systems + microphone setup were more common at that time.

Even though localization is perhaps not as good with speakers as with HRTF and headphones, I feel the environmental audio is much more immersive on speaker systems, which is why I prefer speakers over headphones.

However, such configurations might remain a niche use case.

> So perhaps more productively, @mirh @Hiradur do you have a specific game/speaker setup in mind for using mumble with full physical attenuation/reverb/occlusion/etc?

A few candidates come to my mind:

* Unreal Tournament 2004 (has EAX support and is supported via Mumble Link)
* Battlefield 2 (has EAX support and is supported via Mumble Link)
* Rigs of Rods (has OpenAL EFX and native Mumble support)

I am going to move soon and most likely won't be able to test with the speaker system for a few months since rebuilding it has low priority.
However, if support for OpenAL was implemented, should it work with HRTF over headphones, it should work with speakers as well.

**Comment by mirh** [comment]
> My idea was to add an I3DL2 interface to the Link plugin for game integration so that the current reverb parameters of a game could be set for Mumble via this interface.

There really isn't much of a point doing that, when nonetheless any plugin would still first have to figure where the data resides?
I mean.. I guess that since that (and EAX) are standards probably many games share the same presets, so you could "save" some code by having eax.h be integrated with the program rather than carrying it for each individual plugin, but still it's just a very small thing. A trifle. 

Though as I was yapping in the aforementioned issue, I suppose that it would be possible to have some kind of ~~eyecandy~~ earcandy where playing any supported openal/directsound game could distort with effects to a degree your voice (unclear if an universal plugin would be possible at least in this case where you can control part of the game code, or if you would need a plugin for each one).

> A few candidates come to my mind:

BF2 is funny because project reality ships with its own mumble fork with blackjack and hookers, but it would still probably be a pain.
For whatever the prototyping it would be probably better to have UT2k4 (or that godot demo I linked) as a reference since 1. it might/should have its own kind of internal 3D-aware voice chat to compare 2. its source code is kind of lowkey available
